### PR TITLE
Linked-Cells: Combine SoA Buffer C01

### PR DESCRIPTION
**Pull request by Garfield96**
# Description

Implements an combination of SoA buffers just before the main functor calculation.

- [x] implement algorithm for C01
- [x] make own traversal out of C01 with combined buffers
- [x] add template parameter for OMP loop collapsion
- [x] investigate if combined SoA buffers are also reasonable for C08

# How Has This Been Tested?

- [x] Jenkins

## Note
already includes spdlog v1.3.1

### CODE REVIEW

**Comment by Garfield96** [comment]
Combined SoA buffers are not reasonable for C08, since it would cause a memory overhead, which is characterized by a cubic growth of memory consumption depending on the size of the domain. I'll implement a C04 traversal (other PR) and will try to include combination of SoA buffers there.

**Comment by FG-TUM** [comment]
@Garfield96 ok cool. For the thesis try to explain precisely why c08 is not useful here.

File: src/autopas/utils/SoA.h [hunk 1]
```diff
@@ -129,7 +139,7 @@ class SoA {
    *
    * @return Number of particles.
    */
-  size_t getNumParticles() const { return soaStorage.template get<0>().size(); }
+  inline constexpr size_t getNumParticles() const { return soaStorage.template get<0>().size() - viewStart; }
```
Comment (FG-TUM) [inline src/autopas/utils/SoA.h]: does `constexpr` really do anything here?

File: src/autopas/utils/SoA.h [hunk 2]
```diff
@@ -148,12 +158,18 @@ class SoA {
   }
 
   /**
-   * delete the last particle in the soa
+   * delete the last particle in the soa.
```
Comment (FG-TUM) [inline src/autopas/utils/SoA.h]:
```suggestion
   * Delete the last particle in the SoA.
```

File: src/autopas/containers/linkedCells/traversals/C01Traversal.h [hunk 3]
```diff
@@ -92,64 +100,175 @@ class C01Traversal : public C01BasedTraversal<ParticleCell, PairwiseFunctor, Dat
...
 }
 
-template <class ParticleCell, class PairwiseFunctor, DataLayoutOption DataLayout, bool useNewton3>
-inline void C01Traversal<ParticleCell, PairwiseFunctor, DataLayout, useNewton3>::traverseCellPairs(
+template <class ParticleCell, class PairwiseFunctor, DataLayoutOption DataLayout, bool useNewton3, bool combineSoA>
+inline void C01Traversal<ParticleCell, PairwiseFunctor, DataLayout, useNewton3, combineSoA>::traverseCellPairs(
     std::vector<ParticleCell> &cells) {
   if (not this->isApplicable()) {
     utils::ExceptionHandler::exception(
         "The C01 traversal cannot work with enabled newton3 (unless only one thread is used)!");
```
[the remark this answers is not in the document]
Comment (Garfield96) [reply src/autopas/containers/linkedCells/traversals/C01Traversal.h]: Extended message text in a43497460832452c15bfe80c0acce9b1f990d4e7

File: src/autopas/utils/SoA.h (hunk 1 above)
Comment (Garfield96) [reply src/autopas/utils/SoA.h]: removed in b6494b3d43368c28864d0a02778020c5a8878d30

**Comment by FG-TUM** [comment]
as soon there is a workaround for this wired doxygen bug this can be merged

**Comment by Garfield96** [comment]
> as soon there is a workaround for this wired doxygen bug this can be merged

Fixed

**Pull request merged**